test/framework/src/main/java/org/opensearch/cluster/MockInternalClusterInfoService.java

```diff
@@ -147,4 +150,14 @@ public Long getShardSize(ShardRouting shardRouting) {
     public void setUpdateFrequency(TimeValue updateFrequency) {
         super.setUpdateFrequency(updateFrequency);
     }
+
+    @Override
+    public void addListener(Consumer<ClusterInfo> clusterInfoConsumer) {
+        super.addListener(clusterInfoConsumer);
+        listeners.add(clusterInfoConsumer);
+    }
+
+    public void invokeListeners(final ClusterInfo clusterInfo) {
+        listeners.forEach(listener -> listener.accept(clusterInfo));
+    }
```

Comment:
This doesn't quite seem right, architecturally. Being able to manually invoke listeners for an arbitrary ClusterInfo seems like it violates encapsulation. Also the fact that you're maintaining a parallel list of listeners will potentially make this brittle to changes in the behavior of InternalClusterInfoService. Is there no other way to accomplish what you need?

Reply:
Actually only way to make these test cases determinsitic is to make all the operations inside ```[refresh](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/InternalClusterInfoService.java#L252)``` function of ```InternalClusterInfoService``` synchronous for the test cases flow. Flaky test cases were happening because diskUsages were updated by ```refresh``` function of InternalClusterInfoService asynchronously (I tried increasing refresh interval but it was still flaky). Essentially if you compare the changes I did, it is performing a similar set of operation that is happening inside this ```refresh``` function but in a synchronous way. Also the ClusterInfo object is not random, it is prepared in the same way as it is prepared in the ```refresh``` function (only ```leastAvailableUsagesBuilder``` and ```reservedSpace``` is populated as only these two information is required inside DiskThresoldMonitor.)

One way to make this cleaner will be to override ```addListener``` and ```refresh``` function inside ```MockInternalClusterInfoService``` and make operations inside ```refresh``` function synchronous in ```MockInternalClusterInfoService```. It make sense as well because ```MockInternalClusterInfoService``` is used only  in test cases and asynchronous refresh will cause race condition scenarios in test cases. Let me know your thoughts on this.

Reply:
Looks like we do not need to synchronise refresh function of ```InternalClusterInfoService``` as it refreshes the ClusterInfo in a blocking fashion. Just disabling auto refresh and explicitly calling refresh function after indexing fixed the issue. I have tested it with more than 50 iterations and it is working correctly on local. Let me know if you have any other concern.

Reply:
Looks like synchronising refresh function inside ```InternalClusterInfoService``` will not be required as it refreshes the ClusterInfo in a blocking fashion. Simply disabling auto refresh (auto refresh was making test cases flaky) of ```InternalClusterInfoService``` and explicitly calling refresh after index on node is populated fixed the issue for me (Ignore previous errors as I missed disabling auto refresh on cluster manager node). I have validated test cases with more than 50 iterations on local.

Let me know if you see anything else can cause issue.